### DISCUSSION TITLE
Bagaimana Cara Membuat Custom Route?

### DISCUSSION THREAD

**Discussion by yuliakbard** [mecha-cms/mecha#269] [Q&A] [answered]
misalnya saya buat halaman `/about` ataupun halaman `/trending` di setiap halaman di setiap kategori

seperti ini:
`/article/trending`
`/news/trending`
`/sports/trending`

lebih dari beberapa kategori tapi dengan ditambah diakhir path url `/trending` maka akan terurut dengan sort by view.

**Comment by yuliakbard** [comment]
Cara seperti ini udah benar kah kang? kode nya jalan sih.

isi dalam file `index.php`

```
Hook::set('route', function ($content, $path) {
    if (trim($path, 'trending') . 'trending' === $path) {
        State::set([
            'has' => [
                'next' => true,
                'page' => true,
                'pages' => true,
                'part' => false,
                'prev' => false
            ],
            'is' => [
                'error' => false,
                'page' => true,
                'pages' => false,
                'trend' => true
            ]
        ]);
        type('text/html');
        status(200);
        $GLOBALS['page'] = $page = new Page(null, [
            'title' => 'Trending' ." ". $GLOBALS['t'],
        ]);
$GLOBALS['t'][] = $page->title;
            return layout::include('trend.page');
    }
    return $content;
}, 0);
```

code pada `layout::include('trend.page');` seperti ini:

```
<?php
$pages = Pages::from(LOT . D . 'page' . D . trim($url->path, 'trending') ?? "", 'page', true)->sort([-1, 'view'])->chunk(40,0);
?>
<ul>
<?php foreach($pages as $page): ?>
<li><a href="<?= $page->url ?? $page->link; ?>"><?=$page->title;?></a></li>
    <?php endforeach; ?>
</ul>
```

cuma mainkan di kondisional nya aja kan kang?

`    if (trim($path, 'trending') . 'trending' === $path) {
...
}`

**Comment by taufik-nurrohman** [comment]
Iya.

~~~ php
if ($path && str_ends_with($path, '/trending')) {}
~~~

**Reply by yuliakbard** [reply]
kalau untuk bagian title biar berubah sesuai dengan halamannya gimana ya kang?

```
$GLOBALS['page'] = $page = new Page(null, [
    'title' => 'Trending' ." ". $GLOBALS['t'], // ingin merubah judul sesuai halamannya
]);
$GLOBALS['t'][] = $page->title;
```

**Reply by taufik-nurrohman** [reply]
~~~ php
$GLOBALS['t'][] = $page->title;
$GLOBALS['t'][] = 'Trending';
~~~

**Reply by yuliakbard** [reply]
`$GLOBALS['t'][] = 'Trending';` Kalau pake kode ini judulnya cuma `Trending` aja walau sudah pindah-pindah halaman, kalau bisa judulnya berubah sesuai halamannya gitu kang. 🙏

`/article/trending` => judulnya itu `Trending - Article`
`/news/trending` => judulnya itu `Trending - News`
dll

**Reply by taufik-nurrohman** [reply]
Mungkin karena `$page->title` kosong.

**Reply by yuliakbard** [reply]
ceritanya ane kan ingin mengambil `title` dari `Pages::from(LOT . D . 'page' . D . trim($path, 'trending'), 'page', true);` karena judulnya diset di `$GLOBALS` dengan kode ini.

```
$GLOBALS['page'] = $page = new Page(null, [
    'title' => 'Trending' ."-". $page->title], // Output: Trending - Article
]);
$GLOBALS['t'][] = $page->title;
```

setelah dikulik ane pikir coba ambil 'title' dari file .page nya, misal dengan $path = 'article' dengan kode yang terpikirkan seperti ini:

```
// misal $path itu  'article'
    if ($path && str_ends_with($path, '/trending')) {
        $pages = Pages::from(LOT . D . 'page' . D . trim($path, 'trending'), 'page', true)->sort([1, 'view'])->chunk(20, 0); // akan menampilkan data dari folder  ../page/article
        $page = Page::from(LOT . D . 'page' . D . trim($path, '/trending') . '.page'); // inikan bakal ambil data dari ../page/article.page
        $title = $page->title;
        State::set([
            'has' => [
                'next' => true,
                'page' => true,
                'pages' => true,
                'part' => false,
                'prev' => false,
                'parent' => true
            ],
            'is' => [
                'error' => false,
                'page' => true,
                'pages' => true,
                'trend' => true
            ]
        ]);
        type('text/html');
        status(200);
        $GLOBALS['page'] = $page = new Page(null, [
            'title' => 'Trending' .' '.  $title,
        ]);
        $GLOBALS['t'][] = $page->title;
        return layout::include('trend.page', ['pages' => $pages,'title'=> $title]);
    }
```
apakah cara ini benar kang?

`$page = Page::from(LOT . D . 'page' . D . trim($path, '/trending') . '.page'); // data tidak keluar title yang ingin ane ambil tidak muncul` tetapi kalau `$page = Page::from(LOT . D . 'page' . D . 'article' . '.page') // data title yang mau diambil keluar`

**Reply by taufik-nurrohman** [reply, marked as answer]
Mas salah memahami cara kerja argumen ke dua dari fungsi `trim()`. Kalau mau spesifik ke deretan string tertentu, harusnya pakai `substr()`:

~~~ php
$suffix = '/trending';
if ($path && str_ends_with($path, $suffix)) {
    $path = substr($path, 0, -strlen($suffix));
}
~~~

**Reply by yuliakbard** [reply]
makasih kang. solved.

**Comment by yuliakbard** [comment]
kalau misalnya, ane mau tampilin `$page->content` yang ada pada halaman '/article/1' ke '/article/about' itu gimana ya kang?

**Reply by taufik-nurrohman** [reply]
Coba cek layout [mecha-cms/y.bootstrap.album](https://github.com/mecha-cms/y.bootstrap.album) di berkas `index.php` ada caranya untuk membawa data laman &ldquo;About&rdquo; dan &ldquo;Contact&rdquo;.
